### ISSUE TITLE
Tuples with multiple unsized types shouldn't be valid

### ISSUE THREAD

**Issue by tomaka** [rust-lang/rust#31426]
This code should trigger a compilation error because `[u8]` isn't sized:

``` rust
fn test(t: &([u8], [u16])) -> &[u16] {
    &t.1
}
```

Instead it triggers an LLVM assertion error:

```
rustc: /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/llvm/lib/IR/Instructions.cpp:1304: void llvm::StoreInst::AssertOK(): Assertion `getOperand(0)->getType() == cast<PointerType>(getOperand(1)->getType())->getElementType() && "Ptr must be a pointer to Val type!"' failed.
```

([playpen](http://is.gd/6SgqDh))

**Comment by tomaka** [comment]
I had a debate on IRC whether or not tuples with multiple unsized types shoud be allowed at all.
For the moment they work as long as you don't use instantiate them (they can be useful for metaprogramming purposes). But it's ambiguous whether they are supposed to work.

**Comment by huonw** [comment]
Related to https://github.com/rust-lang/rust/issues/30276.

**Comment by tomaka** [comment]
Also related to https://github.com/rust-lang/rfcs/pull/1592

**Comment by Mark-Simulacrum** [comment]
This no longer ICEs, and gives the following error, which is correct: closing.

```
error[E0277]: the trait bound `[u8]: std::marker::Sized` is not satisfied
 --> test.rs:1:1
  |
1 | / fn test(t: &([u8], [u16])) -> &[u16] {
2 | |     &t.1
3 | | }
  | |_^ the trait `std::marker::Sized` is not implemented for `[u8]`
  |
  = note: `[u8]` does not have a constant size known at compile-time
  = note: tuple elements must have `Sized` type
```
